routing/http/server/filters.go

```diff
@@ -0,0 +1,201 @@
+package server
+
+import (
+	"errors"
+	"reflect"
+	"slices"
+	"strings"
+
+	"github.com/ipfs/boxo/routing/http/types"
+	"github.com/ipfs/boxo/routing/http/types/iter"
+	"github.com/multiformats/go-multiaddr"
+)
+
+// filters implements IPIP-0484
+
+func parseFilter(param string) []string {
+	if param == "" {
+		return nil
+	}
+	return strings.Split(strings.ToLower(param), ",")
+}
+
+// applyFiltersToIter applies the filters to the given iterator and returns a new iterator.
+func applyFiltersToIter(recordsIter iter.ResultIter[types.Record], filterAddrs, filterProtocols []string) iter.ResultIter[types.Record] {
+	mappedIter := iter.Map(recordsIter, func(v iter.Result[types.Record]) iter.Result[types.Record] {
+		if v.Err != nil || v.Val == nil {
+			return v
+		}
+
+		switch v.Val.GetSchema() {
+		case types.SchemaPeer:
+			record, ok := v.Val.(*types.PeerRecord)
+			if !ok {
+				logger.Errorw("problem casting find providers record", "Schema", v.Val.GetSchema(), "Type", reflect.TypeOf(v).String())
+				// TODO: Do we want to let failed type assertions to pass through?
+				return v
+			}
+
+			record = applyFilters(record, filterAddrs, filterProtocols)
+			if record == nil {
+				return iter.Result[types.Record]{Err: errors.New("record is nil")}
```

Comment:
`applyFilters` returns nil when all providers are filtered out (no providers selected by filters), so this does not seem like it should be treated as an error. I think just returning an empty result, or use a defined `ErrNoProviders` error that can be differentiated for actual errors, would be preferable.

Reply:
Good point. I'll return an empty result.